src/charm.py

```diff
@@ -111,274 +69,92 @@ def __init__(self, *args):
-                self.on[f"{self.restart.name}"].acquire_lock.emit()
-
-        # update client_properties whenever possible
-        self.config_manager.set_client_properties()
-
-        # If Kafka is related to client charms, update their information.
-        if self.model.relations.get(REL_NAME, None) and self.unit.is_leader():
-            self.update_client_data()
-
-    def _on_update_status(self, _: EventBase) -> None:
-        """Handler for `update-status` events."""
-        if not self.healthy or not self.upgrade.idle:
-            return
-
-        if not self.state.zookeeper.broker_active():
-            self._set_status(Status.ZK_NOT_CONNECTED)
-            return
-
-        # NOTE for situations like IP change and late integration with rack-awareness charm.
-        # If properties have changed, the broker will restart.
-        self.on.config_changed.emit()
+        self._set_os_config()
 
+    def _set_os_config(self) -> None:
+        """Sets sysctl config."""
         try:
-            if not self.health.machine_configured():
-                self._set_status(Status.SYSCONF_NOT_OPTIMAL)
-                return
-        except SnapError as e:
-            logger.debug(f"Error: {e}")
-            self._set_status(Status.SNAP_NOT_RUNNING)
-            return
-
-        self._set_status(Status.ACTIVE)
+            self.sysctl_config.configure(OS_REQUIREMENTS)
+        except (sysctl.ApplyError, sysctl.ValidationError, sysctl.CommandError) as e:
+            logger.error(f"Error setting values on sysctl: {e.message}")
+            self._set_status(Status.SYSCONF_NOT_POSSIBLE)
 
     def _on_remove(self, _) -> None:
         """Handler for stop."""
         self.sysctl_config.remove()
 
-    def _on_secret_changed(self, event: SecretChangedEvent) -> None:
-        """Handler for `secret_changed` events."""
-        if not event.secret.label or not self.state.cluster.relation:
-            return
+    def _set_status(self, key: Status) -> None:
+        """Sets charm status."""
+        status: StatusBase = key.value.status
+        log_level: DebugLevel = key.value.log_level
+
+        getattr(logger, log_level.lower())(status.message)
+        self.unit.status = status
+
+    def _on_config_changed(self, _):
+        """Handler for `config_changed` events.
 
-        if event.secret.label == self.state.cluster.data_interface._generate_secret_label(
-            PEER,
-            self.state.cluster.relation.id,
-            "extra",  # pyright: ignore[reportArgumentType] -- Changes with the https://github.com/canonical/data-platform-libs/issues/124
+        This handler is in charge of stopping the workloads, since the sub-operators would not
+        be instantiated if roles are changed.
+        """
+        if BROKER.value not in self.config.roles and self.broker.workload.active():
+            self.broker.workload.stop()
+
+        if (
+            BALANCER.value not in self.config.roles
+            and self.unit.is_leader()
```

Comment:
Yes, you are absolutely right. This additional condition is there because I originally added loggers, and it was polluting the logs with unnecessary "Cruise control stopped" messages for units where it was not running in the first place.
I do not remember why I removed them, though. Do you prefer:
- adding the loggers back, and keeping the condition?
- adding the loggers back, and removing the condition?
- only removing the condition?